backend/app/main/service/family_service.py

```diff
@@ -1,68 +1,181 @@
+    return query
+
+def create_papers(paper_query):
+    """Transform the query format to a dictionary format."""
+    papers = db.engine.execute(paper_query)
+    paper_dictionary = defaultdict(list)
+    for paper in papers:
+        from_paper = paper['paper']
+        paper_dictionary[from_paper] = {
             "id": from_paper,
-            "title": connection['from_title'],
-            "abstract": connection['from_abstract'],
-            "year": connection['from_year'],
+            "title": paper['title'],
+            "abstract": paper['abstract'],
+            "year": paper['year'],
+            "cluster": 0,
             "citations": [],
             "authors": [],
         }
-        dictionary[to_paper] = {
-            "id": to_paper,
-            "title": connection['to_title'],
-            "abstract": connection['to_abstract'],
-            "year": connection['to_year'],
-            "citations": [],
-            "authors": []
-        }
+    return paper_dictionary
+
+def add_citations(citations_query, paper_dictionary):
+    """Fill the citation field of every paper in the dictionary."""
+    citations = db.engine.execute(citations_query)
+    for citation in citations:
+        from_paper = citation['from_paper']
+        paper_dictionary[from_paper]['citations'].append(citation['to_paper'])
+    return paper_dictionary
+
+def add_authors(author_query, paper_dictionary):
+    """Fill the author field of every paper in the dictionary."""
+    writes = db.engine.execute(author_query)
+    for write in writes:
+        from_paper = write['paper']
+        paper_dictionary[from_paper]['authors'].append(write['author'])
+    return paper_dictionary
+
+def paper_dictionary_to_paper_list(paper_dictionary):
+    """Transform the paper dictionary into a data list."""
+    paper_list = []
+    for key in paper_dictionary:
+        paper_list.append(paper_dictionary[key])
+    return paper_list
+
+def sort_paper_list(paper_list):
+    """Sort the paper list by paper year."""
+    keyfunc = lambda paper: paper['year']
+    sorted_paper_list = sorted(paper_list, key=keyfunc)
+    return sorted_paper_list
+
+def add_clusters(sorted_paper_list):
+    """Cluster the papers by abstract keywords. Assign each paper its cluster."""
+    if len(sorted_paper_list) == 0:
+        return sorted_paper_list
+
+    abstract_list = extract_abstracts(sorted_paper_list)
+    vectorizer, model = train_clusters(abstract_list)
+    sorted_paper_list = predict_clusters(sorted_paper_list, vectorizer, model)
+    return sorted_paper_list
+
+def extract_abstracts(papers):
+    """Extract an abstract list from the paper list."""
+    abstracts = []
+    for paper in papers:
+        abstracts.append(paper['abstract'])
+    return abstracts
+
+def train_clusters(abstract_list):
+    """Train the clustering algorithm."""
+    vectorizer = TfidfVectorizer(stop_words='english')
```

Comment:
hmm interesting. I didn't realize this before, but now I remembered that we also have some german papers in our DB. Should be noted, but as most of them are english anyway, this is good to go.